### DISCUSSION TITLE
Visualize inputs to custom functions?

### DISCUSSION THREAD

**Discussion by turbanisch** [ropensci/targets#452] [Ideas]
As a newcomer to the targets package I am amazed by its schematic view of the project structure using `tar_visnetwork()`, yet I was wondering why function inputs are not displayed as such. To give an example: Let's say I am using a custom function to merge two datasets and build a new target. Then both datasets and the function that takes them as arguments would appear next to each other, all pointing to the same target. Wouldn't it make sense if the two datasets pointed to the function which in return pointed to the merged result? 

As a newcomer my intuition may well be naive. Perhaps the idea does not lend itself well if the steps to create a specific target include many lines of code rather than just a single custom function (but even then, would it not make sense?). To keep my first pipeline clean, I was indeed just using a single custom function for each transformation between targets like such: 

```
list(
  tar_target(data1, download_data1()),
  tar_target(data2, download_data2()),
  tar_target(data1_cleaned, clean_data1(data1)),
  tar_target(data2_cleaned, clean_data2(data2)),
  tar_target(final, merge_data(data1_cleaned, data2_cleaned), format = "file")
)
```
Any thoughts on the idea (and my habit to outsource all steps into custom functions) are very much appreciated!

**Comment by wlandau** [comment]
`tar_visnetwork()` has a different interpretation because it is a dependency graph. Its goal is to show which targets, functions and objects are directly upstream of any given target. For example, `data1_cleaned` depends on the target `data1` and the function `clean_data1()`. The graph shows a directed edge from `data1` to `data1_cleaned` and another directed edge from `clean_data1()` to `data1_cleaned()` in order to communicate:

1. `clean_data1()` and `data1` need to be processed before `data1_cleaned` is ready to run.
2. If you make changes to either `clean_data1()` or `data1`, then `data1_cleaned` needs to rerun in order to stay up to date. Otherwise, if there are no other changes affecting `data1_cleaned`, then the `data1_cleaned` target can be skipped.

Checking for these dependency relationships, especially looking for disconnected nodes and missing edges, is a useful way to help validate or debug a pipeline.

To represent specific function calls in the graph, we would have to deconstruct each target's command, which may be an expression more complicated than a single function call, and we would need to repeat a bunch of function nodes if the function is called from multiple places. These complicated graphs are produced by https://github.com/moodymudskipper/flow but are less useful to understand the dependency structure of a `targets` pipeline or reveal patterns of target invalidation.

**Comment by turbanisch** [comment]
Thank you so much for this clarification! Having a function to view the dependency graph totally makes sense now. I imagine it would be cool to have another function that does something like the `flow` package to quickly get an overview of the project but very likely it would only work well in simple use cases like mine.
